README.md
# zuo-deploy

基于 Vue3 + Node.js 的轻量 Linux 操作面板，支持简单实时终端、自动化部署、持续集成、Nginx 配置管理/操作、Https 证书管理等。

![version-v1.0.8](https://img.shields.io/badge/version-v1.0.8-yellow.svg) ![license-MIT](https://img.shields.io/badge/license-MIT-green.svg)

**注意：不支持 windows 系统，仅支持 Linux/MacOS 等**

![doc-1-deploy.png](./docImages/doc-1-deploy.png)

## 使用

```bash
# 全局安装
npm install zuo-deploy pm2 -g

# 新建一个目录，用于 zuo-deploy 运行目录，创建的部署脚本会存放到该目录
mkdir zuodeploy; cd zuodeploy;

# 开启服务
zuodeploy start
# ✔ 请指定部署服务监听端口： … 7777
# ✔ 请设置登录密码（默认：888888） … ******

# 访问 127.0.0.1:7777 打开操作界面，密码 888888
# 如果是服务器：服务器ip:7777
```

## 部署脚本管理

支持多项目并行部署，log 互不干扰。新建脚本注意文件名需要以 .sh 命名

![doc-4-deploy-m.png](./docImages/doc-4-deploy-m.png)

### Vue前端项目部署脚本

以下为 visitors.zuo11.com 项目自动化部署脚本（vue3+ts+vite）

```bash
# 进入项目目录
cd /root/visitors

# 获取最近更新
git pull;
git log -1; # 查看最近一次提交 log

# 部署前端服务
cd visitors-fe;
# 如果需要安装新 npm 包
# npm install;
npm run build;
echo '前端服务部署完成'
```
### Node接口服务部署脚本

以下为 api.zuo11.com 接口服务自动化部署脚本

```bash
# 进入 api.zuo11.com 目录
cd /root/api.zuo11.com

git config --global core.quotepath false  # 防止中文乱码
echo "git pull"
git pull
git log -1 # 查看最近一次提交 log

# 如果需要安装新 npm 包
# npm install 

echo '重新开启服务'
pm2 delete api.zuo11.com
pm2 start index.js -n 'api.zuo11.com'
echo '部署完成'
```

### 通用部署脚本
```bash
echo "开始部署..."

# 进入项目目录
cd /var/www/hello-world

# 防止部署 log 中文乱码
git config --global core.quotepath false 

echo "git pull"
git pull 

# 查看最近一次提交 log，了解当前部署的是哪个版本
echo "git log -1"
git log -1 

# 构建相关
# npm install
# 构建
npm run build

echo "部署完成!"
```

### 其他任务类脚本

除了部署，也可以用于数据备份、跑一些脚本任务。

## Nginx/Https证书管理

支持 nginx 配置管理、多 nginx 配置管理、https 证书管理、nginx 服务重启等。

![doc-2-nginx.png](./docImages/doc-2-nginx.png)

### Nginx 配置参考
```bash
# For more information on configuration, see:
#   * Official English Documentation: http://nginx.org/en/docs/
#   * Official Russian Documentation: http://nginx.org/ru/docs/

user root;
worker_processes auto;
error_log /var/log/nginx/error.log;
pid /run/nginx.pid;

# Load dynamic modules. See /usr/share/doc/nginx/README.dynamic.
include /usr/share/nginx/modules/*.conf;

events {
    worker_connections 1024;
}

http {
    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile            on;
    tcp_nopush          on;
    tcp_nodelay         on;
    keepalive_timeout   65;
    types_hash_max_size 4096;

    include             /etc/nginx/mime.types;
    default_type        application/octet-stream;

    # Load modular configuration files from the /etc/nginx/conf.d directory.
    # See http://nginx.org/en/docs/ngx_core_module.html#include
    # for more information.
    include /etc/nginx/conf.d/*.conf;
    
    gzip on;
    gzip_vary on;
    gzip_min_length 1000;
    gzip_comp_level 2;
    gzip_types text/plain text/css text/xml text/javascript application/x-javascript application/xml image/jpeg image/gif image/png application/javascript;

    server {
        listen       80;
        listen       [::]:80;
        server_name  xx.com www.xx.com;
        charset utf-8;

        if ($host = 'xx.com') {
            rewrite ^/(.*)$ http://www.xx.com/$1 permanent;
        }
        
        #charset koi8-r;

        #access_log  logs/host.access.log  main;

        location / {
        #   root   C:\Users\Administrator\Desktop\dist;
           root  /root/xx.com/dist;
           index  index.html index.htm;
        }
        
        # 配置 404
        error_page  404 /root/zuo11.com/dist/404.html;

        # redirect server error pages to the static page /50x.html
        #
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   html;
        }
    }
    
    server {
        listen   80;
        server_name  demo.xx.com;
        charset  utf-8;
        location / {
       	 	root  /root/demo.xx.com;
       	 	index  index.html index.htm;
   	 	}
    }
    
    server {
        listen   80;
        server_name  test.xx.com;
        charset  utf-8;
        location / {
            root   /root/test/test-fe/dist;     
            index index.html index.htm;
            try_files $uri $uri/ /index.html;
   	 	}
    }

    server {
        listen       443 ssl;
        server_name  api.zuo11.com;

        ssl_certificate     cert/3391782_api.xxx.com.pem;
        ssl_certificate_key cert/3391782_api.xx.com.key;

        ssl_session_cache    shared:SSL:1m;
        ssl_session_timeout  5m;

        # ssl_ciphers  HIGH:!aNULL:!MD5;
        # ssl_prefer_server_ciphers  on;

        ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE:ECDH:AES:HIGH:!NULL:!aNULL:!MD5:!ADH:!RC4;  #使用此加密套件。
		ssl_protocols TLSv1 TLSv1.1 TLSv1.2;   #使用该协议进行配置。
		ssl_prefer_server_ciphers on;

        location / {
            # root   html;
            # index  index.html index.htm;
            proxy_pass http://127.0.0.1:7778;
        }
    }
}
```
## 实时终端

支持简单终端，但涉及 vi 等需要等待输入的操作，建议使用 ssh 直连。一般常用功能

- mkdir, cd, ls, pwd
- chmod 0777 xxx（修改目录权限）
- 其他 terminal 操作

![doc-3-terminal.png](./docImages/doc-3-terminal.png)

## 全局设置
- 脚本执行超时时间，默认 2 分钟（推荐），可选 5, 10, 30。主要防止服务器 cpu 占用率过高，减少资源/内存占用。

## 其他
### 服务 log 查询

zuodeploy start 会用 pm2 开启一个 zuodeploy 服务，再次执行 zuodeploy start 会删除原服务，再次开启新服务。**如果开启失败，重新运行一次命令即可**

```bash
# 查看 log
pm2 log
pm2 log zuodeploy --lines 1000 # 指定行
```

### pm2 相关

如果直接使用 node xx.js 运行服务，进程很容易被杀掉，这里使用 pm2 以守护进程方式后台运行服务

```js
pm2 stop zuodeoploy
pm2 start src/index.js -n 'zuodeoploy'
```

跨文件传参, 文件读写

## 历史版本

Tag | 描述 | 实现文档
--- | --- | ---
[v0.3.2](https://github.com/zuoxiaobai/zuo-deploy/tree/v0.3.2) | 用 js 写一个 CI、CD 工具 |  [Vue + Node.js 从 0 到 1 实现自动化部署工具](http://www.zuo11.com/blog/2022/2/zuo_deploy_think.html)

## License

MIT


CHANGELOG.md

# CHANGELOG

## 1.0.8（2022-11-22）

- feat: 新增部署脚本超时时间设置
- feat: 将外部引用 static 文件，切换到个人 cdn
- fix: 修复 child 不能正常终止问题
- docs: 更新 README.md

## 1.0.7（2022-11-08）

zuo-deploy 升级为 Linux 操作面板

- feat: 支持实时终端（轻量）。
- feat: 支持多项目部署、自动化部署脚本管理（增、改、删）。
- feat: 支持 Nginx 配置管理（增、改、删），支持 Nginx 快捷操作。
- feat: 支持 Https 证书文件管理。

## 0.3.2（2022-03-06）

- fix:  修复 log \n 被忽略显示异常的问题，增加推荐的部署 shell 脚本内容

## 0.3.1（2022-02-26）

- feat: 使用 prompt 将端口、密码改为询问方式输入

## 0.3.0（2022-02-26）

- feat: 使用 pm2 改造服务，防止 terminal 中断后进程被杀掉

## 0.2.2（2022-02-11）

- fix: npm 官网 github 链接更正
- fix: socket.io cdn 链接在服务器上运行时异常，将 cdn 替换为官方 cdn

## 0.2.0（2022-02-10）

- feat: 新增 socket 实时输出部署 log
- feat: 启动命令增加端口、密码参数配置
- feat: 引入session, 新增部署接口鉴权、前端登录功能
- feat: 对页面 UI 进行简单修改, 优化提示, 提高 log 展示性能
- feat: 部署 log 加时间戳
- chore: 去掉默认的 ES Module，使用默认的 CommonJS

## 0.1.4（2022-02-08）

- 初步完成自动化部署功能
- koa 提供自动化部署监听服务，请求接口服务，直接开始部署、执行shell脚本。部署完成后在前端页面显示部署 log
- 命令行工具支持 zuodeploy start 开启服务


server/index.js
const Koa = require("koa");
const KoaStatic = require("koa-static");
const KoaRouter = require("koa-router");
const session = require("koa-session");
const bodyParser = require("koa-bodyparser");
const path = require("path");
const fs = require("fs");
const runCmd = require("./utils/runCmd");
const logger = require("./utils/logger");

const app = new Koa();
const router = new KoaRouter();

app.use(bodyParser()); // 处理 post 请求参数

// 参数获取
let argsInfo = fs.readFileSync("args.json").toString();
let args = {};
try {
  args = JSON.parse(argsInfo);
} catch (e) {
  logger.info(e);
}

// 集成 session
app.keys = [`${args.password}`]; // 'some secret hurr'
const CONFIG = {
  key: "koa:sess" /** (string) cookie key (default is koa:sess) */,
  /** (number || 'session') maxAge in ms (default is 1 days) */
  /** 'session' will result in a cookie that expires when session/browser is closed */
  /** Warning: If a session cookie is stolen, this cookie will never expire */
  maxAge: 0.5 * 3600 * 1000, // 0.5h
  overwrite: true /** (boolean) can overwrite or not (default true) */,
  httpOnly: true /** (boolean) httpOnly or not (default true) */,
  signed: true /** (boolean) signed or not (default true) */,
  rolling: false /** (boolean) Force a session identifier cookie to be set on every response. The expiration is reset to the original maxAge, resetting the expiration countdown. (default is false) */,
  renew: false /** (boolean) renew session when session is nearly expired, so we can always keep user logged in. (default is false)*/,
};
app.use(session(CONFIG, app));

// 开启 socket 服务
let socketList = [];
const server = require("http").Server(app.callback());
const socketIo = require("socket.io")(server);
socketIo.on("connection", (socket) => {
  socketList.push(socket);
  logger.info("a user connected");
});

router.get("/isLogin", async (ctx) => {
  ctx.body = {
    code: 0,
    data: !!ctx.session.isLogin,
    msg: "false 未登录，true 登录",
  };
});

router.post("/login", async (ctx) => {
  let code = 0;
  let msg = "登录成功";
  let { password } = ctx.request.body;
  if (password === `${args.password}`) {
    ctx.session.isLogin = true;
  } else {
    code = -1;
    msg = "密码错误";
  }
  ctx.body = {
    code,
    msg,
  };
});

router.post("/deploy", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }

  // 执行部署脚本
  // koa 注意异步 404 的问题
  let execFunc = () => {
    return new Promise((resolve, reject) => {
      try {
        runCmd(
          "sh",
          ["./deploy-master.sh"],
          function (text) {
            resolve(text);
          },
          socketIo,
          "deploy-log"
        );
      } catch (e) {
        logger.info(e);
        reject(e);
      }
    });
  };

  try {
    let res = await execFunc();
    ctx.body = {
      code: 0,
      msg: res,
    };
  } catch (e) {
    ctx.body = {
      code: -1,
      msg: e.message,
    };
  }
});

router.post("/runShell", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }

  let { shellText, shellName = "temp.sh", timeoutMinute = 2 } = ctx.request.body;
  const shellFilePath = process.cwd() + "/" + shellName;
  console.log("shellFilePath", shellFilePath);
  try {
    fs.writeFileSync(shellFilePath, shellText);
  } catch (e) {
    console.log(e);
    ctx.body = {
      code: -1,
      msg: e.message,
    };
  }

  // // 执行部署脚本
  // // koa 注意异步 404 的问题
  let execFunc = () => {
    return new Promise((resolve, reject) => {
      try {
        runCmd(
          "sh",
          [shellFilePath],
          function (text, isError) {
            isError ? reject(new Error(text)) : resolve(text);
          },
          socketIo,
          "runShell",
          timeoutMinute
        );
      } catch (e) {
        logger.info(e);
        reject(e);
      }
    });
  };

  try {
    let res = await execFunc();
    ctx.body = {
      code: 0,
      msg: res,
    };
  } catch (e) {
    ctx.body = {
      code: -1,
      msg: e.message,
    };
  }
});

router.get("/shell/get", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }

  const files = fs.readdirSync(`${process.cwd()}`);
  let result = [];
  // files [ 'args.json', 'temp.sh', 'test.sh' ]
  let filesFilter = files.filter((item) => {
    if (["args.json", "temp.sh"].includes(item)) {
      return false;
    }
    if (item.startsWith(".")) {
      return false;
    }
    return true;
  });
  console.log(filesFilter);

  filesFilter.forEach((filePath) => {
    console.log(`'${filePath}'`);
    let info = { name: filePath, content: "", desc: "" };
    try {
      const content = fs
        .readFileSync(`${process.cwd()}/${filePath}`)
        .toString();
      info.content = content;
      result.push(info);
    } catch (e) {
      console.log("非文件", e.message);
    }
  });

  // [
  //   { name: "1.sh", content: "ls\npwd\n", desc: "定时任务" },
  //   {
  //     name: "2.sh",
  //     content: "git pull;\n npm i && npm run build\n",
  //     desc: "项目部署",
  //   },
  // ],
  ctx.body = {
    code: 0,
    data: {
      processCwd: process.cwd(),
      shellList: result,
    },
    msg: "成功",
  };
});

router.post("/runCurShell", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }

  let { name, timeoutMinute } = ctx.request.body;
  const shellFilePath = process.cwd() + "/" + name;
  console.log("cur shellFilePath", shellFilePath);

  // // 执行部署脚本
  // // koa 注意异步 404 的问题
  let execFunc = () => {
    return new Promise((resolve, reject) => {
      try {
        runCmd(
          "sh",
          [shellFilePath],
          function (text, isError) {
            isError ? reject(new Error(text)) : resolve(text);
          },
          socketIo,
          "shell-log-" + name,
          timeoutMinute
        );
      } catch (e) {
        logger.info(e);
        reject(e);
      }
    });
  };

  try {
    let res = await execFunc();
    ctx.body = {
      code: 0,
      msg: res,
    };
  } catch (e) {
    ctx.body = {
      code: -1,
      msg: e.message,
    };
  }
});

router.post("/saveFile", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { name, content, curPath } = ctx.request.body;
  console.log(name, content, curPath);
  !curPath && (curPath = ".");
  let prePath = curPath.startsWith("/")
    ? curPath
    : path.join(process.cwd(), curPath);

  const filePath = prePath + (prePath.endsWith("/") ? "" : "/") + name;
  console.log("filePath", filePath, prePath);
  try {
    fs.writeFileSync(filePath, content);
    ctx.body = {
      code: 0,
      msg: "成功",
    };
  } catch (e) {
    console.log(e);
    ctx.body = {
      code: -1,
      msg: `文件 '${filePath}' 创建失败，` + e.message,
    };
  }
});

router.post("/editFile", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { name, content, curPath } = ctx.request.body;
  // console.log(name, content, curPath);
  !curPath && (curPath = ".");
  let prePath = curPath.startsWith("/")
    ? curPath
    : path.join(process.cwd(), curPath);
  const filePath = prePath + (prePath.endsWith("/") ? "" : "/") + name;
  // console.log("filePath", filePath, prePath);
  try {
    fs.writeFileSync(filePath, content);
    ctx.body = {
      code: 0,
      msg: "成功",
    };
  } catch (e) {
    console.log(e);
    ctx.body = {
      code: -1,
      msg: `文件 '${filePath}' 修改失败，` + e.message,
    };
  }
});

router.post("/deleteFile", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { name, curPath } = ctx.request.body;
  console.log(name, curPath);
  !curPath && (curPath = ".");
  let prePath = curPath.startsWith("/")
    ? curPath
    : path.join(process.cwd(), curPath);
  const filePath = prePath + (prePath.endsWith("/") ? "" : "/") + name;
  // console.log("filePath", filePath, prePath);
  try {
    fs.rmSync(filePath);
    ctx.body = {
      code: 0,
      msg: "成功",
    };
  } catch (e) {
    console.log(e);
    ctx.body = {
      code: -1,
      msg: `文件 '${filePath}' 删除失败，` + e.message,
    };
  }
});

router.get("/nginx/get", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { curPath } = ctx.query;
  console.log(">>", curPath);

  try {
    const content = fs.readFileSync(curPath).toString();
    ctx.body = {
      code: 0,
      data: content,
      msg: "成功",
    };
  } catch (e) {
    ctx.body = {
      code: -3,
      msg: e.message,
    };
  }
});

router.get("/nginx/multipleGet", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { curPath } = ctx.query;
  console.log(">>", curPath);

  let files = [];
  try {
    files = fs.readdirSync(curPath);
  } catch (e) {
    ctx.body = {
      code: -3,
      msg: e.message,
    };
    return;
  }

  let result = [];
  // files [ 'args.json', 'temp.sh', 'test.sh' ]
  let filesFilter = files.filter((item) => {
    // if ([].includes(item)) {
    //   return false;
    // }
    if (item.startsWith(".")) {
      return false;
    }
    return true;
  });
  console.log(filesFilter);

  filesFilter.forEach((filePath) => {
    console.log(`'${filePath}'`);
    let info = { name: filePath, content: "", desc: "" };
    try {
      const content = fs.readFileSync(`${curPath}/${filePath}`).toString();
      info.content = content;
      result.push(info);
    } catch (e) {
      console.log("非文件", e.message);
    }
  });
  ctx.body = {
    code: 0,
    data: result,
    msg: "成功",
  };
});

router.get("/https/get", async (ctx) => {
  if (!ctx.session.isLogin) {
    ctx.body = {
      code: -2,
      msg: "未登录",
    };
    return;
  }
  let { curPath } = ctx.query;
  console.log(">>", curPath);

  let files = [];
  let finalPath = curPath + (curPath.endsWith("/") ? "" : "/") + "cert";
  let isExist = true;
  try {
    fs.statSync(finalPath);
  } catch (e) {
    // stats 失败 Error: ENOENT: no such file or directory, stat '/etc/nginx/cert'
    // 文件不存在，创建该目录
    isExist = false;
  }

  try {
    !isExist && fs.mkdirSync(finalPath);
    files = fs.readdirSync(finalPath);
  } catch (e) {
    ctx.body = {
      code: -3,
      msg: e.message,
    };
    return;
  }

  let result = [];
  // files [ 'args.json', 'temp.sh', 'test.sh' ]
  let filesFilter = files.filter((item) => {
    // if ([].includes(item)) {
    //   return false;
    // }
    if (item.startsWith(".")) {
      return false;
    }
    return true;
  });
  console.log(filesFilter);

  filesFilter.forEach((filePath) => {
    console.log(`'${filePath}'`);
    let info = { name: filePath, content: "", desc: "" };
    const content = fs.statSync(`${finalPath}/${filePath}`);
    info.content = content;
    result.push(info);
  });
  ctx.body = {
    code: 0,
    data: result,
    msg: "成功",
  };
});

const multer = require("@koa/multer");
// 文件上传处理
// 前端 append 文件时使用的是 image 字段
let singleFileConfig = multer().single("image");
let multipleFilesConfig = multer().fields([
  {
    name: "image",
    maxCount: 5,
  },
]);
let isMultiple = true;
let fileConfig = isMultiple ? multipleFilesConfig : singleFileConfig;
router.post("/upload", fileConfig, async (ctx) => {
  // 文件外的其他 FormData数据 { param1: 'abc' }
  console.log("ctx.request.body", ctx.request.body);
  console.log("ctx.files", ctx.files); // 多文件，返回 { 字段1: [file数组], 字段2: [file数组] }
  console.log("ctx.file", ctx.file); // 单文件，返回 file 对象
  let { curPath } = ctx.request.body;
  let finalPath = curPath + (curPath.endsWith("/") ? "" : "/");

  // 如果是单文件取文件内容，如果是多文件，取第一个文件，前端字段传的 image
  // 在服务端本地创建文件
  try {
    if (isMultiple) {
      let files = ctx.files["image"];
      files.forEach((item) => {
        let { originalname, buffer } = item;
        fs.writeFileSync(finalPath + originalname, buffer);
      });
    } else {
      let { originalname, buffer } = ctx.file;
      fs.writeFileSync(finalPath + originalname, buffer);
    }
    ctx.body = {
      code: 0,
      msg: "成功",
      data: ctx.request.body,
    };
  } catch (e) {
    ctx.body = {
      code: -4,
      msg: e.message,
    };
  }
  // {
  //   fieldname: 'image',
  //   originalname: '截屏2020-12-10 下午8.01.44.png',
  //   encoding: '7bit',
  //   mimetype: 'image/png',
  //   buffer: <Buffer 89 50 4e 47 0d 0a 1a 0a 00 00 00 0d 49 48 44 52 00 00 03 18 00 00 01 56 08 06 00 00 00 ea b0 3b 51 00 00 0c 64 69 43 43 50 49 43 43 20 50 72 6f 66 69 ... 90135 more bytes>,
  //   size: 90185
  // }
});

app.use(new KoaStatic(path.resolve(__dirname, "../frontend")));
app.use(router.routes()).use(router.allowedMethods());
server.listen(args.port, () => logger.info(`服务监听 ${args.port} 端口`));
